Auto-generate operator CRD docs

ptgott opened the pull request:
Closes #32165 
Closes #34023 

We already pull JSON schema data from Teleport protobuf message definitions in order to generate CRD manifests. Extract a shared library from code within the existing protoc plugin for generating CRDs. This way, we can use the library to generate docs for our CRDs with a separate plugin, and don't need to complicate the interface for using the existing plugin.

- Organize `integrations/crdgen` into `lib` and `cmd` directories, turning `lib` into a package.
- Edit the `generateSchema` signature to include a function that formats a CRD into a document, injecting separate functions for formatting the CRD as YAML and as a docs page.
- Add Makefile targets for building the docs plugin and generating CRD docs.

To display the resulting reference pages in the docs, this change configures `docs/config.json` to generate the Reference section of the sidebar from the docs directory. This requires moving the Helm reference includes directory into `docs/pages/includes`.

ptgott committed:
Auto-generate operator CRD docs

We already pull JSON schema data from Teleport protobuf message
definitions in order to generate CRD manifests. Extract a shared library
from code within the existing protoc plugin for generating CRDs. This
way, we can use the library to generate docs for our CRDs with a
separate plugin, and don't need to complicate the interface for using
the existing plugin.

- Organize `integrations/crdgen` into `lib` and `cmd` directories,
  turning `lib` into a package.
- Edit the `generateSchema` signature to include a function that formats
  a CRD into a document, injecting separate functions for formatting the
  CRD as YAML and as a docs page.
- Add Makefile targets for building the docs plugin and generating CRD
  docs.

To display the resulting reference pages in the docs, this change
configures `docs/config.json` to generate the Reference section of the
sidebar from the docs directory. This requires moving the Helm reference
includes directory into `docs/pages/includes`.

hugoShaka commented on docs/pages/reference/operator-resources/resources.teleport.dev_rolesv7.mdx:
```diff
@@ -0,0 +1,402 @@
+|create_db_user_mode||CreateDatabaseUserMode allows users to be automatically created on a database when not set to off. 0 is "unspecified", 1 is "off", 2 is "keep", 3 is "best_effort_drop".|
+|create_desktop_user|boolean|CreateDesktopUser allows users to be automatically created on a Windows desktop|
+|create_host_user|boolean|CreateHostUser allows users to be automatically created on a host|
+|create_host_user_mode||CreateHostUserMode allows users to be automatically created on a host when not set to off. 0 is "unspecified"; 1 is "off"; 2 is "drop" (removed for v15 and above), 3 is "keep"; 4 is "insecure-drop".|
```
Enum types are missing. I'm not sure what we accept here, at least the integers, maybe the string representation. @tigrato already fought with enums in CRDs

hugoShaka reviewed (commented):
Looks very good 🎉

ptgott commented on docs/pages/reference/operator-resources/resources.teleport.dev_accesslists.mdx:
```diff
@@ -0,0 +1,102 @@
+This guide is a comprehensive reference to the fields available to configure in
+the `TeleportAccessList` resource, which you can apply after installing
+the Teleport Kubernetes operator.
+
```
That sounds great! What do you think would be the best way to include examples? Map each example file to a CRD in `formatAsDocsPage` based on the example filename?

hugoShaka replied:
We could have formatAsDocsPage lookup a yaml file named like the CRD and include it if found.

ptgott replied:
Sounds good! Since this is a Q2 goal for us, would it make sense to merge the docs generator this quarter, then work on the examples as time opens up?

ptgott committed:
Respond to hugoShaka feedback

- Do not display the `status` field.
- Indicate that enum fields can take an integer or a string. This is
  consistent with our comment conventions, which describe both the
  integer and the string values for each enum option.

ptgott committed:
Sort CRD reference page sections

Also handle duplicate page sections by including the names of parent
properties within section headings.

ptgott commented:
TODO: need to remove files from this PR that were touched without changes

hugoShaka commented:
This PR breaks the crd generation as the whole `cmd/protoc-gen-crd` is gone. Did you forget to commit a file?

```bash
$ cd integrations/operator
$ make crdgen
make -C crdgen
cd cmd/protoc-gen-crd; go build -o protoc-gen-crd
/bin/sh: line 0: cd: cmd/protoc-gen-crd: No such file or directory
no Go files in /Users/shaka/go/src/github.com/gravitational/teleport/integrations/operator/crdgen
make[1]: *** [build] Error 1
make: *** [crdgen] Error 2
```

ptgott commented:
@hugoShaka Thanks for catching that! The `protoc-gen-crd` directory happened to match a `.gitignore` value, so the directory only existed on my local. I've fixed the `.gitignore` and tested on a fresh clone

hugoShaka committed:
Fix crdgen Makefile targets

- bring test data and hack out of `lib/`
- fix broken debug builds
- add crd-docs checks to the `crds-up-to-date` target
- trigger proto lint on CRD and CRD doc changes
- removed stale gitkeep in `docs/pages/includes/crd-docs/.gitkeep`
- split crdgen formatting logic in another file for readability
- update license header years on new files

hugoShaka committed:
fixup! Fix crdgen Makefile targets

hugoShaka committed:
lint files due to false positive 'generated' comment

hugoShaka committed:
fixup! lint files due to false positive 'generated' comment

ptgott committed:
Incorporate rosstimothy feedback

Remove an unnecessary package.

Also regenerate the CRD docs.

ptgott commented on integrations/operator/crdgen/lib/debug.go:
Thanks! ✅

hugoShaka replied:
Can you also update the debug binaries (both `debug.go`) so they compile?

ptgott replied:
Hi Hugo, I hadn't realized there was a debug version, but now I see the README. I'll get that fixed. Was there supposed to be a `debug.go` in each subdirectory of `cmd`? It looks like the `.gitignore` in `crdgen` may be leaving them out of the branch.

hugoShaka replied:
I thought I commited the debug files 🤔  I'll take a look. Maybe some gitignore is too zealous, or I just forgot

hugoShaka committed:
fix debug builds and gitignore

hugoShaka committed:
fixup! fix debug builds and gitignore

ptgott committed:
Respond to rosstimothy feedback

Fix integrations/operator/crdgen/Makefile

ptgott commented on integrations/operator/crdgen/cmd/protoc-gen-crd-docs/debug.go:
```diff
@@ -0,0 +1,58 @@
+	"os"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
```
Since the original scope of this PR was to make the minimum set of changes required to format CRD data into documentation pages, would it make sense to use `slog` in a separate PR? The original CRD generator plugin uses `logrus`, and it doesn't seem like `slog` is a drop-in replacement.

ptgott committed:
Fix spelling

ptgott commented:
Some tests are failing intermittently—I'm going to look into this before merging!

ptgott committed:
Fix flaky propertyTable test

Compare the unsorted slice elements, since sorting takes place on the
containing VersionSection.
